File: src/mesh_generator.py
import os

import gmsh
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.lines import Line2D
from matplotlib.patches import Polygon, Rectangle
from matplotlib.collections import PatchCollection


class MeshGenerator:
    """A class to generate 2D meshes using gmsh."""

    def __init__(self, surface_tags, output_dir="."):
        """
        Initializes the Mesh2D class.

        Args:
            surface_tags (list or int): A list of surface tags or a single surface tag.
            output_dir (str): The directory to save the output files.
        """
        if isinstance(surface_tags, int):
            self.surface_tags = [surface_tags]
        else:
            self.surface_tags = surface_tags
        self.output_dir = output_dir
        os.makedirs(self.output_dir, exist_ok=True)

    def generate(
        self,
        mesh_params: dict,
        filename="mesh.msh",
        show_nodes=False,
        show_cells=False,
    ):
        """
        Generates the mesh.

        Args:
            mesh_params (dict): A dictionary where keys are surface tags and values are dicts
                                with 'mesh_type' and 'char_length'.
            filename (str): The path to save the output .msh file.
        """
        for surface_tag in self.surface_tags:
            if surface_tag not in mesh_params:
                continue

            params = mesh_params[surface_tag]
            mesh_type = params.get("mesh_type", "tri")

            if mesh_type not in ["structured", "tri", "quads"]:
                raise ValueError("mesh_type must be 'structured', 'tri', or 'quads'")

            if mesh_type == "structured":
                char_length = params.get("char_length", 0.1)
                # For structured mesh, we need to define transfinite properties
                # on all the new curves and surfaces.
                gmsh.model.mesh.setTransfiniteSurface(surface_tag)
                gmsh.model.mesh.setRecombine(2, surface_tag)

                boundary_curves = gmsh.model.getBoundary(
                    [(2, surface_tag)], oriented=False
                )

                if len(boundary_curves) != 4:
                    raise ValueError(
                        "Structured mesh is only supported for geometries with 4 boundary curves."
                    )

                # Determine nx and ny from char_length
                bbox = gmsh.model.getBoundingBox(2, surface_tag)
                dx = bbox[3] - bbox[0]
                dy = bbox[4] - bbox[1]
                nx = int(dx / char_length)
                ny = int(dy / char_length)

                horizontal_curves = []
                vertical_curves = []

                for curve_dim_tag in boundary_curves:
                    curve_tag = curve_dim_tag[1]
                    p_tags = gmsh.model.getBoundary([curve_dim_tag], oriented=False)
                    p_start_tag = p_tags[0][1]
                    p_end_tag = p_tags[1][1]
                    coord_start = gmsh.model.occ.getCenterOfMass(0, p_start_tag)
                    coord_end = gmsh.model.occ.getCenterOfMass(0, p_end_tag)

                    if abs(coord_start[1] - coord_end[1]) < 1e-6:  # Horizontal
                        horizontal_curves.append(curve_tag)
                    else:  # Vertical
                        vertical_curves.append(curve_tag)

                for curve_tag in horizontal_curves:
                    gmsh.model.mesh.setTransfiniteCurve(curve_tag, nx + 1)

                for curve_tag in vertical_curves:
                    gmsh.model.mesh.setTransfiniteCurve(curve_tag, ny + 1)

            elif mesh_type == "quads":
                gmsh.model.mesh.setRecombine(2, surface_tag)

        # --- Physical Groups ---
        all_boundary_curves = []
        for surface_tag in self.surface_tags:
            boundary_curves = gmsh.model.getBoundary([(2, surface_tag)], oriented=False)
            all_boundary_curves.extend([c[1] for c in boundary_curves])

        gmsh.model.addPhysicalGroup(1, list(set(all_boundary_curves)), name="boundary")
        gmsh.model.addPhysicalGroup(2, self.surface_tags, name="fluid")

        # Generate the 2D mesh
        gmsh.model.mesh.generate(2)

        # Save the mesh
        msh_file = os.path.join(self.output_dir, filename)
        gmsh.write(msh_file)
        print(f"Successfully created mesh.")
        print(f"Mesh saved to: {msh_file}")

        # Plot the mesh
        self.plot(
            file_name=filename.replace(".msh", ".png"),
            show_nodes=show_nodes,
            show_cells=show_cells,
        )

    def plot(self, file_name="mesh.png", show_nodes=False, show_cells=False):
        """Plots the generated mesh with options to show cell and node labels."""
        fig, ax = plt.subplots(figsize=(12, 10))

        # Get nodes
        node_tags, node_coords, _ = gmsh.model.mesh.getNodes()
        x = node_coords[0::3]
        y = node_coords[1::3]
        node_map = {tag: i for i, tag in enumerate(node_tags)}
        num_nodes = len(node_tags)

        # Get cells
        elem_tags_global = []
        elem_node_tags_global = []
        elem_types_global = []
        for surface_tag in self.surface_tags:
            elem_types, elem_tags, elem_node_tags = gmsh.model.mesh.getElements(
                2, surface_tag
            )
            for i, elem_type in enumerate(elem_types):
                elem_types_global.append(elem_type)
                elem_tags_global.append(elem_tags[i])
                elem_node_tags_global.append(elem_node_tags[i])

        num_cells = sum(len(tags) for tags in elem_tags_global)

        # Determine dynamic font size
        base_font_size_cell = 10
        base_font_size_node = 8
        cell_font_scale = (
            max(0.5, 1 - np.log10(num_cells + 1) / 4) if num_cells > 0 else 1
        )
        node_font_scale = (
            max(0.5, 1 - np.log10(num_nodes + 1) / 4) if num_nodes > 0 else 1
        )
        cell_fontsize = base_font_size_cell * cell_font_scale
        node_fontsize = base_font_size_node * node_font_scale

        patches = []
        cell_counter = 0
        for i, elem_type in enumerate(elem_types_global):
            num_nodes_per_elem = 0
            if elem_type == 2:  # 3-node triangle
                num_nodes_per_elem = 3
                color = "#87CEEB"
            elif elem_type == 3:  # 4-node quad
                num_nodes_per_elem = 4
                color = "#90EE90"
            else:
                color = "#FFD700"

            if num_nodes_per_elem > 0:
                num_elem = len(elem_tags_global[i])
                for j in range(num_elem):
                    node_tags_for_elem = elem_node_tags_global[i][
                        j * num_nodes_per_elem : (j + 1) * num_nodes_per_elem
                    ]
                    node_indices = [node_map[tag] for tag in node_tags_for_elem]
                    points = np.array([[x[k], y[k]] for k in node_indices])
                    polygon = Polygon(
                        points, facecolor=color, edgecolor="k", alpha=0.7, lw=0.5
                    )
                    patches.append(polygon)

                    if show_cells:
                        cell_centroid_x = float(np.mean(points[:, 0]))
                        cell_centroid_y = float(np.mean(points[:, 1]))
                        ax.text(
                            cell_centroid_x,
                            cell_centroid_y,
                            str(cell_counter),
                            color="black",
                            ha="center",
                            va="center",
                            fontsize=cell_fontsize,
                            weight="bold",
                            bbox=dict(
                                facecolor="white",
                                alpha=0.6,
                                edgecolor="none",
                                boxstyle="round,pad=0.2",
                            ),
                        )
                    cell_counter += 1

        if show_nodes:
            for i in range(num_nodes):
                ax.text(
                    float(x[i]),
                    float(y[i]),
                    str(i),
                    color="darkred",
                    ha="center",
                    va="center",
                    fontsize=node_fontsize,
                    bbox=dict(
                        facecolor="yellow",
                        alpha=0.6,
                        edgecolor="none",
                        boxstyle="round,pad=0.1",
                    ),
                )

        p = PatchCollection(patches, match_original=True)
        ax.add_collection(p)

        model_name = gmsh.model.getCurrent() or "default"
        ax.set_title(f"{model_name} Mesh", fontsize=16, weight="bold", pad=20)
        ax.set_xlabel("X-coordinate", fontsize=12)
        ax.set_ylabel("Y-coordinate", fontsize=12)
        ax.grid(True, linestyle="--", alpha=0.5)
        ax.set_aspect("equal", adjustable="box")
        ax.autoscale_view()

        for spine in ax.spines.values():
            spine.set_visible(False)

        legend_handles = [
            Rectangle((0, 0), 1, 1, color="#87CEEB", label="Triangle"),
            Rectangle((0, 0), 1, 1, color="#90EE90", label="Quads"),
            Rectangle((0, 0), 1, 1, color="#FFD700", label="Other"),
        ]

        ax.legend(
            handles=legend_handles,
            loc="upper right",
            bbox_to_anchor=(1, 1.025),
            fontsize=10,
            frameon=False,
            ncol=len(legend_handles),
        )

        plot_file = os.path.join(self.output_dir, file_name)
        plt.savefig(plot_file, dpi=300, bbox_inches="tight")
        plt.close()
        print(f"Mesh plot saved to: {plot_file}")

File: src/mesh_generator.cpp
#include "mesh_generator.hpp"
#include <gmsh.h>
#include <algorithm>
#include <cmath>
#include <filesystem>
#include <iostream>
#include <set>
#include <stdexcept>
#include <unordered_set>

namespace fvm {

MeshGenerator::MeshGenerator(const std::vector<int>& surfaceTags,
                             const std::string& outputDir)
    : surfaceTags_(surfaceTags), outputDir_(outputDir) {
    ensureOutputDir();
}

MeshGenerator::MeshGenerator(int surfaceTag, const std::string& outputDir)
    : surfaceTags_({surfaceTag}), outputDir_(outputDir) {
    ensureOutputDir();
}

void MeshGenerator::ensureOutputDir() {
    std::filesystem::create_directories(outputDir_);
}

void MeshGenerator::generate(const std::map<int, MeshParams>& meshParams,
                             const std::string& filename) {
    // Apply mesh parameters for each surface
    for (int surfaceTag : surfaceTags_) {
        auto it = meshParams.find(surfaceTag);
        if (it != meshParams.end()) {
            applyMeshParameters(surfaceTag, it->second);
        }
    }

    // Set up physical groups
    setupPhysicalGroups();

    // Generate 2D mesh
    gmsh::model::mesh::generate(2);

    // Save mesh to Gmsh format
    saveMesh(filename);

    // Extract mesh data for further processing
    extractMeshData();

    std::cout << "Mesh generation complete." << std::endl;
    std::cout << "  Nodes: " << meshData_.nodes.size() << std::endl;
    std::cout << "  Cells: " << meshData_.cells.size() << std::endl;
}

void MeshGenerator::applyMeshParameters(int surfaceTag, const MeshParams& params) {
    switch (params.meshType) {
        case MeshType::Structured:
            setStructuredMesh(surfaceTag, params.charLength);
            break;
        case MeshType::Quads:
            gmsh::model::mesh::setRecombine(2, surfaceTag);
            break;
        case MeshType::Triangles:
            // Default meshing, no special settings needed
            break;
    }
}

void MeshGenerator::setStructuredMesh(int surfaceTag, double charLength) {
    gmsh::model::mesh::setTransfiniteSurface(surfaceTag);
    gmsh::model::mesh::setRecombine(2, surfaceTag);

    // Get boundary curves
    std::vector<std::pair<int, int>> boundaryCurves;
    gmsh::model::getBoundary({{2, surfaceTag}}, boundaryCurves, false, false, false);

    if (boundaryCurves.size() != 4) {
        throw std::runtime_error(
            "Structured mesh is only supported for geometries with 4 boundary curves.");
    }

    // Get bounding box for calculating number of divisions
    double minX, minY, minZ, maxX, maxY, maxZ;
    gmsh::model::getBoundingBox(2, surfaceTag, minX, minY, minZ, maxX, maxY, maxZ);

    double dx = maxX - minX;
    double dy = maxY - minY;
    int nx = static_cast<int>(dx / charLength);
    int ny = static_cast<int>(dy / charLength);

    // Classify curves and set transfinite
    auto [hCurves, vCurves] = classifyBoundaryCurves(boundaryCurves);

    for (int curveTag : hCurves) {
        gmsh::model::mesh::setTransfiniteCurve(curveTag, nx + 1);
    }
    for (int curveTag : vCurves) {
        gmsh::model::mesh::setTransfiniteCurve(curveTag, ny + 1);
    }
}

std::pair<std::vector<int>, std::vector<int>>
MeshGenerator::classifyBoundaryCurves(
    const std::vector<std::pair<int, int>>& boundaryCurves) const {

    std::vector<int> hCurves, vCurves;

    for (const auto& dimTag : boundaryCurves) {
        int curveTag = dimTag.second;

        // Get curve endpoints
        std::vector<std::pair<int, int>> pointTags;
        gmsh::model::getBoundary({{1, curveTag}}, pointTags, false, false, false);

        if (pointTags.size() < 2) continue;

        int pStartTag = pointTags[0].second;
        int pEndTag = pointTags[1].second;

        std::vector<double> coordStart, coordEnd;
        std::vector<double> paramCoord;  // unused but required
        gmsh::model::getValue(0, pStartTag, paramCoord, coordStart);
        gmsh::model::getValue(0, pEndTag, paramCoord, coordEnd);

        // Classify by orientation
        if (std::abs(coordStart[1] - coordEnd[1]) < 1e-6) {
            hCurves.push_back(curveTag);
        } else {
            vCurves.push_back(curveTag);
        }
    }

    return {hCurves, vCurves};
}

void MeshGenerator::setupPhysicalGroups() {
    // Collect all boundary curves
    std::set<int> allBoundaryCurves;
    for (int surfaceTag : surfaceTags_) {
        std::vector<std::pair<int, int>> boundary;
        gmsh::model::getBoundary({{2, surfaceTag}}, boundary, false, false, false);
        for (const auto& dimTag : boundary) {
            allBoundaryCurves.insert(dimTag.second);
        }
    }

    // Find curves already in physical groups
    std::set<int> curvesInGroups;
    std::vector<std::pair<int, int>> physicalGroups;
    gmsh::model::getPhysicalGroups(physicalGroups, 1);

    for (const auto& [dim, tag] : physicalGroups) {
        std::vector<int> entities;
        gmsh::model::getEntitiesForPhysicalGroup(dim, tag, entities);
        for (int e : entities) {
            curvesInGroups.insert(e);
        }
    }

    // Add untagged curves to "unnamed" group
    std::vector<int> untaggedCurves;
    for (int curve : allBoundaryCurves) {
        if (curvesInGroups.find(curve) == curvesInGroups.end()) {
            untaggedCurves.push_back(curve);
        }
    }
    if (!untaggedCurves.empty()) {
        gmsh::model::addPhysicalGroup(1, untaggedCurves, -1, "unnamed");
    }

    // Handle 2D physical groups (surfaces)
    std::set<int> surfacesInGroups;
    gmsh::model::getPhysicalGroups(physicalGroups, 2);

    for (const auto& [dim, tag] : physicalGroups) {
        std::vector<int> entities;
        gmsh::model::getEntitiesForPhysicalGroup(dim, tag, entities);
        for (int e : entities) {
            surfacesInGroups.insert(e);
        }
    }

    // Add untagged surfaces to "fluid" group
    std::vector<int> untaggedSurfaces;
    for (int surface : surfaceTags_) {
        if (surfacesInGroups.find(surface) == surfacesInGroups.end()) {
            untaggedSurfaces.push_back(surface);
        }
    }
    if (!untaggedSurfaces.empty()) {
        gmsh::model::addPhysicalGroup(2, untaggedSurfaces, -1, "fluid");
    }
}

void MeshGenerator::saveMesh(const std::string& filename) {
    std::string mshFile = outputDir_ + "/" + filename;
    gmsh::write(mshFile);
    std::cout << "Mesh saved to: " << mshFile << std::endl;
}

void MeshGenerator::extractMeshData() {
    extractNodes();
    extractCells();
    extractPhysicalGroups();
    extractFaces();
}

void MeshGenerator::extractNodes() {
    std::vector<std::size_t> nodeTags;
    std::vector<double> nodeCoords;
    std::vector<double> parametricCoords;

    gmsh::model::mesh::getNodes(nodeTags, nodeCoords, parametricCoords);

    meshData_.nodes.clear();
    meshData_.nodeIds.clear();
    meshData_.nodes.reserve(nodeTags.size());
    meshData_.nodeIds.reserve(nodeTags.size());

    for (std::size_t i = 0; i < nodeTags.size(); ++i) {
        meshData_.nodeIds.push_back(nodeTags[i]);
        meshData_.nodes.push_back({
            nodeCoords[3 * i],
            nodeCoords[3 * i + 1],
            nodeCoords[3 * i + 2]
        });
    }
}

void MeshGenerator::extractCells() {
    // Create node tag to index map
    std::unordered_map<std::size_t, std::size_t> nodeMap;
    for (std::size_t i = 0; i < meshData_.nodeIds.size(); ++i) {
        nodeMap[meshData_.nodeIds[i]] = i;
    }

    meshData_.cells.clear();
    meshData_.cellTypes.clear();

    for (int surfaceTag : surfaceTags_) {
        std::vector<int> elemTypes;
        std::vector<std::vector<std::size_t>> elemTags;
        std::vector<std::vector<std::size_t>> elemNodeTags;

        gmsh::model::mesh::getElements(elemTypes, elemTags, elemNodeTags, 2, surfaceTag);

        for (std::size_t i = 0; i < elemTypes.size(); ++i) {
            int elemType = elemTypes[i];

            // Get element properties
            std::string name;
            int dim, order, numNodes, numPrimaryNodes;
            std::vector<double> localNodeCoords;
            gmsh::model::mesh::getElementProperties(
                elemType, name, dim, order, numNodes, localNodeCoords, numPrimaryNodes);

            std::size_t numElements = elemTags[i].size();
            const auto& allNodeTags = elemNodeTags[i];

            for (std::size_t j = 0; j < numElements; ++j) {
                CellConnectivity cell;
                cell.reserve(numNodes);

                for (int k = 0; k < numNodes; ++k) {
                    std::size_t nodeTag = allNodeTags[j * numNodes + k];
                    cell.push_back(nodeMap[nodeTag]);
                }

                meshData_.cells.push_back(std::move(cell));
                meshData_.cellTypes.push_back(getVTKCellType(numNodes));
            }
        }
    }
}

void MeshGenerator::extractPhysicalGroups() {
    meshData_.boundaryGroups.clear();
    meshData_.volumeGroups.clear();

    std::vector<std::pair<int, int>> physicalGroups;
    gmsh::model::getPhysicalGroups(physicalGroups);

    for (const auto& [dim, tag] : physicalGroups) {
        std::string name;
        gmsh::model::getPhysicalName(dim, tag, name);

        std::vector<int> entities;
        gmsh::model::getEntitiesForPhysicalGroup(dim, tag, entities);

        PhysicalGroup group;
        group.dimension = dim;
        group.tag = tag;
        group.name = name.empty() ? "group_" + std::to_string(tag) : name;
        group.entities = entities;

        if (dim == 1) {
            meshData_.boundaryGroups[group.name] = group;
        } else if (dim == 2) {
            meshData_.volumeGroups[group.name] = group;
        }
    }
}

void MeshGenerator::extractFaces() {
    // For 2D meshes, faces are edges
    // Build edge-to-cell connectivity

    // Create node tag to index map
    std::unordered_map<std::size_t, std::size_t> nodeMap;
    for (std::size_t i = 0; i < meshData_.nodeIds.size(); ++i) {
        nodeMap[meshData_.nodeIds[i]] = i;
    }

    // Map from edge (ordered pair) to cells sharing it
    std::map<std::pair<std::size_t, std::size_t>, std::vector<std::size_t>> edgeToCells;

    for (std::size_t cellIdx = 0; cellIdx < meshData_.cells.size(); ++cellIdx) {
        const auto& cell = meshData_.cells[cellIdx];
        std::size_t n = cell.size();

        for (std::size_t i = 0; i < n; ++i) {
            std::size_t n1 = cell[i];
            std::size_t n2 = cell[(i + 1) % n];

            // Create ordered edge key
            auto edge = std::minmax(n1, n2);
            edgeToCells[edge].push_back(cellIdx);
        }
    }

    meshData_.internalFaces.clear();
    meshData_.boundaryFaces.clear();
    meshData_.boundaryFaceLabels.clear();

    // Get boundary elements for labeling
    std::map<std::pair<std::size_t, std::size_t>, std::string> boundaryEdgeLabels;

    for (const auto& [name, group] : meshData_.boundaryGroups) {
        for (int entityTag : group.entities) {
            std::vector<int> elemTypes;
            std::vector<std::vector<std::size_t>> elemTags;
            std::vector<std::vector<std::size_t>> elemNodeTags;

            gmsh::model::mesh::getElements(elemTypes, elemTags, elemNodeTags, 1, entityTag);

            for (std::size_t i = 0; i < elemTypes.size(); ++i) {
                // Line elements have 2 nodes
                const auto& nodeTags = elemNodeTags[i];
                for (std::size_t j = 0; j + 1 < nodeTags.size(); j += 2) {
                    std::size_t n1 = nodeMap[nodeTags[j]];
                    std::size_t n2 = nodeMap[nodeTags[j + 1]];
                    auto edge = std::minmax(n1, n2);
                    boundaryEdgeLabels[edge] = name;
                }
            }
        }
    }

    // Classify edges
    for (const auto& [edge, cells] : edgeToCells) {
        if (cells.size() == 2) {
            // Internal face
            meshData_.internalFaces.push_back({edge.first, edge.second});
        } else if (cells.size() == 1) {
            // Boundary face
            meshData_.boundaryFaces.push_back({edge.first, edge.second});

            auto it = boundaryEdgeLabels.find(edge);
            if (it != boundaryEdgeLabels.end()) {
                meshData_.boundaryFaceLabels.push_back(it->second);
            } else {
                meshData_.boundaryFaceLabels.push_back("unnamed");
            }
        }
    }
}

}  // namespace fvm
